DM-37921: Initial draft

jonathansick opened the pull request:
This is the initial draft of the vertical monorepo architecture for client-server development. https://sqr-075.lsst.io/v/DM-37921/index.html

A companion PR that prototypes the monorepo architecture in safirdemo is https://github.com/lsst-sqre/safirdemo/pull/193

jonathansick committed:
Migrate to beta of new Rubin Sphinx technotes

jonathansick committed:
Ignore tox and .venv from Sphinx build

This should be added to the Sphinx configuration.

jonathansick committed:
Update bibliographies

jonathansick committed:
Write down the problem statement

jonathansick committed:
Describe architectural options

This covers our mini-co-work discussion of where to put the shared
Pydantic models.

jonathansick committed:
Describe the monorepo file layout

jonathansick committed:
Draft notes on the client dependency

I'm not sure if this will work; needs to be tested.

jonathansick committed:
Shorten exploration of alternatives

This summarizes the longer discussion of alternative client library
repositories. Russ also suggested highlighting the argument that it's
hard to develop an app when the client library needs to be
simultaneously developed.

jonathansick committed:
Document dependency handling

This is based on the experience of setting up the Tox environment and
Docker build for safirdemo.

jonathansick committed:
Write about repo structure

jonathansick committed:
Discuss naming strategies for clients

jonathansick committed:
Write about Pydantic model architecture

jonathansick committed:
Add example code based on noteburst

jonathansick committed:
Discuss a new Pydantic utils library

jonathansick committed:
Start describing a sans-I/O arch for clients

jonathansick committed:
Edits

athornton reviewed (approved):
This is great.  I'll have a few minor comments, but nothing at all major.

athornton commented on index.rst:
```diff
@@ -1,29 +1,413 @@
+In practice, the client library's absence from ``requirements/main.txt`` itself isn't harmful because the client is inherently pinned by virtue of being co-developed in the same Git repository.
+
+What's potentially concerning, though, is the absence of the client's own dependencies from the application's ``requirements/main.txt`` dependencies.
+We could mitigate this risk by limiting client library dependencies to packages that are already in the main application's ``requirements/main.txt``.
```
Maybe something to the effect of "we don't expect this to be onerous to developers because client models are generally subsets of richer server models."

athornton commented on index.rst:
```diff
@@ -1,29 +1,413 @@
+   run:
+   	cd server && tox run -e=run
+
+Again, the Docker build uses the ``main.hashed.txt`` requirements, while the Tox environment uses the unhashed ``main.txt`` and ``dev.txt`` files.
```
Clever.

athornton commented on index.rst:
```diff
@@ -1,29 +1,413 @@
+---------------------
+
+We may want to systematically prefix the package names for discovery and sorting on PyPI and Conda-Forge.
+For example, rather than ``noteburst-client``, we may prefer to use ``rsp-notebust-client`` or even ``rubin-rsp-noteburst-client``.
```
"rsp-noteburst-client"

athornton commented on index.rst:
```diff
@@ -1,29 +1,413 @@
+
+This approach future-proofs the client library for new HTTP libraries, and makes the client more widely useful.
+As well, a mock version of the client can be implemented that doesn't do any network requests, but does capture information for introspection.
+Such a mock can be useful for testing.
```
Yes, this!  I think it might be worth thinking about and maybe adding a section.  There's a little tension here: the client doesn't need unit tests, but having standard mocks as part of the client package would make consumption of the client easier, since you could just use the supplied mocks in the code that consumed the client.  I mentioned this a little in the jupyterlabcontroller discussion I had with Russ, but I think we somehow want to roll mocks in with the clients.

jonathansick replied:
Good idea.

jonathansick committed:
Set vscode tasks to run in tox

Setup VS Code tasks (in .vscode/tasks.json) to run two tasks via the tox
environment: compile HTML, and refresh the bib files.

jonathansick committed:
Refresh bib files

jonathansick committed:
Talk about client dependency pinning more

jonathansick committed:
Add conclusion with flowchart

jonathansick committed:
Pin mermaid version for compat with Sphinx ext.
